internal/zinc-core/src/main/scala/sbt/internal/inc/InvalidationProfiler.scala

```diff
@@ -0,0 +1,267 @@
+    def registerInitial(changes: InitialChanges): Unit = {
+      import scala.collection.JavaConverters._
+      val fileChanges = changes.internalSrc
+      val profChanges = zprof.Changes(
+        added = toStringTableIndices(toPathStrings(fileChanges.getAdded.asScala)).toList,
+        removed = toStringTableIndices(toPathStrings(fileChanges.getRemoved.asScala)).toList,
+        modified = toStringTableIndices(toPathStrings(fileChanges.getChanged.asScala)).toList
+      )
+      initial = Some(
+        zprof.InitialChanges(
+          changes = Some(profChanges),
+          removedProducts = toStringTableIndices(toPathStrings(changes.removedProducts)).toList,
+          binaryDependencies = toStringTableIndices(toPathStrings(changes.binaryDeps)).toList,
+          externalChanges = toApiChanges(changes.external).toList
+        )
+      )
+    }
+
+    private final var currentEvents: List[zprof.InvalidationEvent] = Nil
+    def registerEvent(
+        kind: String,
+        inputs: Iterable[String],
+        outputs: Iterable[String],
+        reason: String
+    ): Unit = {
+      val event = zprof.InvalidationEvent(
+        kind = kind,
+        inputs = toStringTableIndices(inputs).toList,
+        outputs = toStringTableIndices(outputs).toList,
+        reason = reason
+      )
+
+      currentEvents = event :: currentEvents
+    }
+
+    private final var cycles: List[zprof.CycleInvalidation] = Nil
+    def registerCycle(
+        invalidatedClasses: Iterable[String],
+        invalidatedPackageObjects: Iterable[String],
+        initialSources: Iterable[File],
+        invalidatedSources: Iterable[File],
+        recompiledClasses: Iterable[String],
+        changesAfterRecompilation: APIChanges,
+        nextInvalidations: Iterable[String],
+        shouldCompileIncrementally: Boolean
+    ): Unit = {
+      val newCycle = zprof.CycleInvalidation(
+        invalidated = toStringTableIndices(invalidatedClasses).toList,
+        invalidatedByPackageObjects = toStringTableIndices(invalidatedPackageObjects).toList,
+        initialSources = toStringTableIndices(toPathStrings(initialSources)).toList,
+        invalidatedSources = toStringTableIndices(toPathStrings(invalidatedSources)).toList,
+        recompiledClasses = toStringTableIndices(recompiledClasses).toList,
+        changesAfterRecompilation = toApiChanges(changesAfterRecompilation).toList,
+        nextInvalidations = toStringTableIndices(nextInvalidations).toList,
+        startTimeNanos = compilationStartNanos,
+        compilationDurationNanos = compilationDurationNanos,
+        events = currentEvents,
+        shouldCompileIncrementally = shouldCompileIncrementally
+      )
+
+      cycles = newCycle :: cycles
+      ()
+    }
+  }
+}
+
+abstract class RunProfiler {
+  def timeCompilation(
+      startNanos: Long,
+      durationNanos: Long
+  ): Unit
+
+  def registerInitial(
+      changes: InitialChanges
+  ): Unit
+
+  def registerEvent(
+      kind: String,
```

Comment:
What `kind`s are allowed here? Is there any other implementation I can look at to see what
is applicable here?

Reply:
It's in generic on purpose so that we can easily involve it in the future, even though we only have one implementation. The kinds are defined in https://github.com/sbt/zinc/pull/585/files#diff-de86f02542b47ffddbb8687a94316daeR259